clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -1756,9 +1798,41 @@ scheduler_impl_tmpl_t<RecordType, ReaderType>::inject_kernel_sequence(
     // odd to have regular records with the new tid before the top headers.
     if (sequence.empty())
         return stream_status_t::STATUS_EOF;
+    bool saw_any_instr = false;
+    bool set_branch_target_marker = false;
+    trace_marker_type_t marker_type;
+    uintptr_t marker_value = 0;
     for (int i = static_cast<int>(sequence.size()) - 1; i >= 0; --i) {
         RecordType record = sequence[i];
         record_type_set_tid(record, input->tid);
+        if (record_type_is_instr(record)) {
+            set_branch_target_marker = false;
+            if (!saw_any_instr) {
+                saw_any_instr = true;
+                // If the last to-be-injected instruction is an indirect branch, set its
+                // indirect_branch_target field to the fallthrough pc of the last
+                // returned instruction from this input (for syscall injection, it would
+                // be the syscall for which we're injecting the trace). This is simpler
+                // than trying to get the actual next instruction on this input for which
+                // we would need to read-ahead.
+                // XXX i#6495, i#7157: The above strategy does not work for syscalls that
+                // transfer control (like sigreturn), but we do not trace those anyway
+                // today (neither using Intel-PT, nor QEMU) as there are challenges in
+                // determining the post-syscall resumption point.
+                if (record_type_is_indirect_branch_instr(record,
+                                                         input->last_pc_fallthrough)) {
+                    // For memref_t we do not see a separate branch_target marker, so
+                    // the following does not have any affect and will simply be
+                    // reset at the next instr.
```

Comment:
Hmm, this, combined with record_type_is_indirect_branch_instr wanting different behavior and the else branch below not applying to memref_t, should all this code inside this loop be moved into type-specific instantiations where it may be easier to understand?

Reply:
I wanted to avoid that code duplication, which I considered significant. Maybe I can have record_type_is_indirect_branch_instr also set a by-reference bool arg for `branch_target_field_set`, so the caller can take action based on that.

Reply:
Or actually, the caller can just make that determination based on whether the branch_target_field is equal to input->last_pc_fallthrough or not.

Reply:
Went ahead with the first one (because the second one is actually not possible since we cannot look into the record in a non-specialized function).